build: Minor build system fixes and amendments

hebasto opened the pull request:
This PR addresses the following comments:
- https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1742342524
- https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1728692369
- https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1736110362
- https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1742931121
- https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1747723657
- https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1742328675
- https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1723106474

hebasto committed:
build, test: Add missed log options

hebasto committed:
build: Print `CMAKE_CXX_COMPILER_ARG1` in summary

When `-DCMAKE_CXX_COMPILER='clang++;-stdlib=libc++;-m32'` is provided,
`-stdlib=libc++ -m32` flags are printed in the summary now.

maflcko reviewed (approved):
review ACK c393acadf213ddf58a69759eaf341595ccbf4889

hebasto commented on CMakeLists.txt:
```diff
@@ -146,6 +146,8 @@ if(WITH_ZMQ)
     # The ZeroMQ project has provided config files since v4.2.2.
     # TODO: Switch to find_package(ZeroMQ) at some point in the future.
+    # Note: As of 2024, mainstream distributions do not yet provide
+    # CMake config files for ZeroMQ packages.
```
Thanks! Reworked per your feedback.

fanquake commented on CMakeLists.txt:
```diff
@@ -145,7 +145,9 @@ if(WITH_ZMQ)
   else()
     # The ZeroMQ project has provided config files since v4.2.2.
-    # TODO: Switch to find_package(ZeroMQ) at some point in the future.
+    # However, as of 2024, mainstream distributions do not yet provide
```
I think we should just remove any dates entirely.

hebasto replied:
Thanks! Done.

hebasto committed:
doc: Amend comment about ZeroMQ config files

hebasto committed:
doc: Drop `ctest` command from Windows cross-compiling instructions

The ctest command was added hastily without considering the requirement
of Wine, which is generally not trivial to install.

hebasto commented:
> Could include [#30454 (comment)](https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1742931121) to avoid having to open another pull for that?

Sure! Included.

fanquake commented:
Let's also wait for https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1743326932. To clarify if this ever worked (and was later broken?), or was just never expected to work. In either case, it could be worth documenting.

fanquake commented on CMakeLists.txt:
```diff
@@ -145,7 +145,9 @@ if(WITH_ZMQ)
   else()
     # The ZeroMQ project has provided config files since v4.2.2.
-    # TODO: Switch to find_package(ZeroMQ) at some point in the future.
+    # However, mainstream distributions do not yet provide CMake
```
It's disappointing that so many distros don't support CMake in their packaging. We basically can't switch this (and any other packages where we might want to use the "native" CMake config) until *every* distro does, otherwise it'll just be broken for that distro.

Some do, i.e [Alpine](https://pkgs.alpinelinux.org/contents?branch=edge&name=zeromq%2ddev&arch=x86&repo=main), or [Arch](https://archlinux.org/packages/extra/x86_64/zeromq/), or [nix](https://github.com/NixOS/nixpkgs/blob/release-23.11/pkgs/development/libraries/zeromq/4.x.nix), but the majority do not, ie Debian, Ubuntu, Fedora, *BSDs, as well as other package managers, i.e `brew` or `Guix`. Gentoo has had an issue open for adding CMake support for ~5 years: https://bugs.gentoo.org/695202, but no change.

There are multiple issues here (and it's not really all the fault of the distros):
* One is that distros won't switch build systems if the current one works fine, and switching doesn't provide (m)any benefits, while having a 100% chance of breaking dependant packages/other things.
* Another is that packages like zeromq, which have opted to provide two build systems, don't properly support installing the config files from one, in the other. 
  * i.e if you build with CMake you'll get CMake config files, and pkg-config files. However if you build with Autotools, you just get pkg-config files.
  * One solution to try and speed up Cmake adoption wouuld be to upstream support for installing CMake config files from Autotools.

I assume one thing we could do, is try `find_package(ZeroMQ)`  first, and fallback to `pkg_check_modules(libzmq)`, and just maintain this (pretty much) forever. That might be better than somewhat vauge comments about "modern" distros, because, (some) modern distros do already support this, and it's actually the case that *all* versions of *all* distros Bitcoin Core supports need to do this, before we could do a complete switch.

Sjors replied:
This explanation makes more sense to me than the current code comment.

> try `find_package(ZeroMQ)` first, and fallback to `pkg_check_modules(libzmq)`

That would be nice. Then a short clarifying comment could be:

```
# Use CMake config file if the distro provides it (pseudo code)
try find_package(ZeroMQ)
# Otherwise 
catch pkg_check_modules(libzmq)
```

fanquake replied:
Discussed offline, and we will change the code to try both in a follow up.

fanquake replied:
Opened an issue for this #30876.

Sjors commented on doc/build-windows.md:
```diff
@@ -67,7 +67,6 @@ Build using:
     gmake -C depends HOST=x86_64-w64-mingw32  # Use "-j N" for N parallel jobs.
     cmake -B build --toolchain depends/x86_64-w64-mingw32/toolchain.cmake
     cmake --build build     # Use "-j N" for N parallel jobs.
-    ctest --test-dir build  # Use "-j N" for N parallel tests. Some tests are disabled if Python 3 is not available.
```
cc9d65dc05f802aaaa108a2edcf30c3758c3f459: maybe add a comment for why there's not test command (yet?)

hebasto committed:
build: Stop enabling CMake's CMP0141 policy

The `CMAKE_MSVC_DEBUG_INFORMATION_FORMAT` variable has not been used
since the merge of https://github.com/hebasto/bitcoin/pull/215 in the
CMake staging branch.

hebasto committed:
build: Delete MSVC special case for `BUILD_FOR_FUZZING` option

hebasto committed:
build: Delete dead code that implements `IF_CHECK_FAILED` option

hebasto commented:
> lgtm ACK [cc9d65d](https://github.com/bitcoin/bitcoin/commit/cc9d65dc05f802aaaa108a2edcf30c3758c3f459)
> 
> Let's wait for the outcome of the discussions in [#30454 (comment)](https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1726684296) and [#30454 (comment)](https://github.com/bitcoin/bitcoin/pull/30454#discussion_r1723106474) as well?

Thanks! Both comments have been addressed.

sipsorcery commented:
tACK 1cc93fe7b40f10a7d1d1189058af98a2bce31381 (win11 msvc).
